File: doc-examples/example-java/src/test/java/example/ExplicitExecutorsTest.java
package example;

import io.micronaut.el.CompiledELContext;
import io.micronaut.el.CompiledExpressionFactory;
import io.micronaut.el.ContributedELMethodExecutor;
import io.micronaut.el.interpreter.InterpretingELExpressionParser;
import jakarta.el.ELContext;
import jakarta.el.ExpressionFactory;
import jakarta.el.MethodNotFoundException;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ExplicitExecutorsTest {

    @Test
    void onlyTheContributedMethodsAreReachable() {
        ExpressionFactory factory = new CompiledExpressionFactory(List.of(), // <1>
            new InterpretingELExpressionParser(List.of(new ContributedELMethodExecutor(List.of(new BookMethods()))))); // <2>
        ELContext context = new CompiledELContext().setBean("book", new Book("Jakarta EL", "reference", 20d));

        assertEquals("Jakarta EL", factory.createValueExpression(context, "${book.title()}", Object.class).getValue(context)); // <3>
        assertThrows(MethodNotFoundException.class,
            () -> factory.createValueExpression(context, "${book.getTitle()}", Object.class).getValue(context)); // <4>
    }
}

File: doc-examples/example-kotlin/src/test/kotlin/example/ExplicitExecutorsTest.kt
package example

import io.micronaut.el.CompiledELContext
import io.micronaut.el.CompiledExpressionFactory
import io.micronaut.el.ContributedELMethodExecutor
import io.micronaut.el.interpreter.InterpretingELExpressionParser
import jakarta.el.MethodNotFoundException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

class ExplicitExecutorsTest {

    @Test
    fun onlyTheContributedMethodsAreReachable() {
        val factory = CompiledExpressionFactory(listOf(), // <1>
            InterpretingELExpressionParser(listOf(ContributedELMethodExecutor(listOf(BookMethods()))))) // <2>
        val context = CompiledELContext().setBean("book", Book("Jakarta EL", "reference", 20.0))

        assertEquals("Jakarta EL", factory.createValueExpression(context, "\${book.title()}", Any::class.java).getValue(context)) // <3>
        assertThrows(MethodNotFoundException::class.java) {
            factory.createValueExpression(context, "\${book.getTitle()}", Any::class.java).getValue(context) // <4>
        }
    }
}
